### DISCUSSION TITLE
Denoising with PCA

### DISCUSSION THREAD

**Discussion by samcoveney** [dipy/dipy#2678] [1. Ideas & Feedback]
Following on from recent issues #2666 and #2655 I believe that we need to look deeper into the current PCA denoising algorithms, which may not be correct.

I suggest that we create a synthetic example in which we generate noise with spatially-varying (but spatially-correlated) variance (we can use both Gaussian and Rician noise) and demonstrate whether the current algorithms (1) estimate the noise level properly; (2) succeed in denoising the images. A synthetic example is simple to generate we can fit an existing dataset and take the fitted results as the ground truth, then add our own noise to this. While the noise level can be spatially correlated, it's probably best to make the noise itself uncorrelated as this is what the algorithms are designed to handle.

Any ideas here are welcome. Happy to lead on this if that helps.

**Comment by skoudoro** [comment]
Thank you for pointing @samcoveney.  Yes, it will be great if you could lead this. 

Also, I think we should organize a developer meeting to talk about it. I will contact some them and keep you updated when I have 2-3dates.

Thank you for your help.

**Reply by samcoveney** [reply]
That's great, please let me know when.

I will be able to start work on this soon. I will probably make a new example that demonstrates results (vs known truth) with the original code, and with new code, so that we can verify what is happening.

**Comment by samcoveney** [comment]
# Samples vs Features
 
Firstly, to introduce some confusion, we have

Veerart 2016 "Diffusion MRI noise mapping using random matrix theory"
> Let X be an M × N real-valued measurement matrix, with rows representing M voxels within a sliding window, and columns representing N weighted magnitude MR measurements in each voxel.

Veerart 2016 "Denoising of diffusion MRI using random matrix theory"
> In what follows, X will be a real-valued dMRI matrix, with rows representing M dMRI measurements and columns representing N voxels within a local neighborhood, typically a sliding window (note that M and N are re-denoted compared to Veraart et al., 2016).

i.e. a change of notation. But either way, the eigenvectors represent the signal over the diffusion directions (i.e. over the different images/acquisitions), so the point is to exploit redundancy in the acquisition/image dimension. In this case, voxels = samples.

Manjon 2013 paper "Diffusion Weighted Image Denoising Using Overcomplete Local PCA" 
> In our proposed approach, we apply a local PCA to exploit the multi-directional redundancy of DWI patterns rather than using local spatial image pattern redundancy

so again, voxels = samples, _except for the estimation of the noise level itself_ in which the method given in the paper says
> we perform a PCA decomposition of a matrix formed by as many columns as image voxels and as many rows as images, see Fig. 2

meaning that the function `pca_noise_estimate` is wrong (not the method from the paper cited in this function's documentation) as discussed in #2655.


# localPCA vs MPPCA

Both `mppca` and `localpca` utilize `genpca`, work by using rows = samples = voxels. There are several issues here:
* `localpca` should estimate the variance using the method in Manjon 2013 paper "Diffusion Weighted Image Denoising Using Overcomplete Local PCA" i.e. `pca_noise_estimate` (which needs fixing, see above), but if `sigma` is not provided, it is estimated with random matrix theory i.e. the method of mppca. Now, this may be fine, perhaps using `localpca` as an overdetermined `mppca` algorithm is desired, but this should at least raise a warning, since as far as I can tell there is a mash-up of methods here.
* following on from above, and discussed in #2666, the `mppca` algorithm uses the overdetermined ideas from `localpca` so it doesn't look like the algorithm from Veerart 2016 "Denoising of diffusion MRI using random matrix theory" is implemented at all (since the result is always combined with the overcomplete idea from Manjon 2013 paper "Diffusion Weighted Image Denoising Using Overcomplete Local PCA". This doesn't seem right, as it's one of the main differences of the two methods


# Possible Errors

Issue #2666 suggests that there is an error in the estimation of the noise level.

In the code we have:

https://github.com/dipy/dipy/blob/d0401e0d8f4234997f2a85a70782ddec005b12b5/dipy/denoise/localpca.py#L43-L51

Looking at Veerart 2016 "Denoising of diffusion MRI using random matrix theory" we have

![image](https://user-images.githubusercontent.com/19683485/206265688-4792e71e-e376-494f-9b51-f44f69522030.png)

So that we need to determine p such that we separate the signal from the noise. Slightly confusing is that the theory is presented with descending eigenvalues, but the code is implemented with ascending eigenvalues, but never mind.

By my calculation, issue #2666 is _incorrect_ in suggesting that there is an error here. With a little rearranging, we can see that the variance that is calculated and returned is correct, and so is the calculation of r for the stopping criterion. It is just that the 'final' form of the variance is used throughout i.e. that of equation (12), rather then equation (11), and the iteration criterion is expressed in terms of this final form of the variance. *But it would be really useful for someone else to take another look please*

**Comment by RafaelNH** [comment]
Hi @samcoveney. I confirm what you've wrote in **sample vs features**. Regarding issue #2655 - before solving it, perhaps it will be useful to fist generalize the genpca function to handle matrices with M<N. We have two solutions for this: 1) transposing matrix X in this condition; or 2) just removing the expected number of zero eigenvalues in vector d (line 232/ line 241) when M<N.

Regarding what you wrote in "**localPCA vs MPPCA**":
1) thumbs up in raising a warning when sigma is not given as a input in the function localPCA.
2) indeed, Veraart et al. 2016 did not use the overcomplete averaging. But, I don't see the point of not including this already in our MPPCA implementations, given that this approach was proven to be more advantageous than just taking the signals on the middle of the sliding window (Katkovnik V, Foi A, Egiazarian K, Astola J (2010) From local kernel to nonlocal Multiple-Model image denoising. Int. J. Comput. Vision, 86(1):1–32). Note that the MPPCA denoising algorithm that we have in DIPY is actually the reference implementation of the denoising algorithm used for my PhD thesis, where the gains provided by overcomplete averaging in the MP-PCA denoising algorithm are assessed (see section 3.4.2 Overcomplete PCA Denoising SNR Gain (page 60 in [https://www.repository.cam.ac.uk/handle/1810/281993](https://www.repository.cam.ac.uk/handle/1810/281993)). 

Regarding what you wrote in "**Possible Errors**":

As you well pointed, the differences between the implemented algorithm and the method session of Veraart et al., 2016 do not correspond to an actual error. You can easily rewrite the equation in Veraart et al., 2016 to match the code in function _pca_classifier. The real reason why the code is different from the paper's description, is that our code was implemented before the time that Veraart's paper was published. I've started my implementations based on Veraart's presentation at the ISMRM 2015, where not all the details of the moment matching procedure was given ([https://www.ismrm.org/15/program_files/FriSci05.htm](https://www.ismrm.org/15/program_files/FriSci05.htm)) - so, instead of writing the expressions in reference to MP distribution variance, I decided to write then in reference to MP distribution bandwidth. 

Another difference between our code and Veraart's description, it that we use the variance estimate instead of p. I did that modification just to make the code more similar to the localpca code, and to better move all redundant parts of both localpca and mppca to the genpca function (see the discussion in #1917). 

With a little refactoring, I think we can rewrite the code exactly as Jelle describes in his manuscript (personally, this is not my top priority because these modifications will not improve the performance of the code).

**Reply by samcoveney** [reply]
Thanks @RafaelNH I will respond in sections below

On **Samples vs Features**

I don't believe that transposing matrix X is a good idea, since this then turns the voxels into features, and then the algorithms really would be quite different? While this would remove the zero-eigenvalues (for cases where they would appear before the transpose), I don't think this problem is equivalent.

Using M: directions N: voxels, before my pull request #2579 the code enforced that M < N meaning that there should not be any non-zero eigenvalues (actually, since the mean is subtracted in `genpca`, if M = N, there would be a single zero eigenvalue... I suppose we should account for losing the degree of freedom from subtracting the mean, or we can just stop subtracting the mean). With #2579 we can have M > N, which means we will have zero-eigenvalues (again, one more than we might think, because we subtracted the mean) but I don't think I have properly accounted for the variance calculation in `_pca_classifier` because...

https://github.com/dipy/dipy/blob/d0401e0d8f4234997f2a85a70782ddec005b12b5/dipy/denoise/localpca.py#L47

...the mean is used, therefore we are summing a significant number of zero eigenvalues but still dividing by a number that counts those eigenvalues. Instead of taking the mean of only the non-zero eigenvalues (again, with possible correction needed for mean subtraction). 

I believe this is exactly what you were indicating in your comment above? I can add this correction to the code.

On issue #2655 do you agree that the current implementation is not correct, and voxels need to be used as features? In this case, we would require more voxels than directions, to make sure there was non-zero variance in the last eigenvalue (once again, we need to account for the mean subtraction here). In the original implementation, the _entire image_ is used rather than patches, so this num_voxels > num_directions is easily satisfied. In the current implementation, which I think has samples and features the wrong way around, the image is processed in patches, so if the number of voxels=samples in a patch is above the number of images=features, the last eigenvalue will be non-zero, which is fine. But I don't think this is the original algorithm. And the current algorithm will also fail if num_voxels (samples) < num_directions (features).

On **localPCA vs MPPCA**

I agree that there seems no point in removing the over-completeness aspect of MPPCA, but we should just make sure to document this. So both algorithms will be over-complete, and we just need to highlight what the differences really are. Can you clarify what you think the differences really are in the current implementation?

On **Possible Errors**

Also, I agree that there seems to be little point in rewriting the `_pca_classifier`, we could just put a comment in there to avoid confusion.

(note: we probably need to correct for M >= N, see above)

When you say "it that we use the variance estimate instead of p" do you mean that, instead of thresholding based on p, we threshold based on the variance, like so:

https://github.com/dipy/dipy/blob/d0401e0d8f4234997f2a85a70782ddec005b12b5/dipy/denoise/localpca.py#L251-L255

Actually, thresholding is based on the corrected variance tau, which I assume is an implementation of

![image](https://user-images.githubusercontent.com/19683485/206429262-8027ffcf-5083-498b-befa-627118ccff64.png)

All this seems fine, just wanted to double check.

Thanks again.

**Reply by samcoveney** [reply]
Okay, I have corrected `_pca_classifier` to work when the number of samples is less than the number of features. I've made some tests and will add these, then I will correct `pca_noise_estimate` (and maybe make it the default choice for "local PCA" if sigma is not provided explicitly). All will be pushed in the new year.

**Comment by samcoveney** [comment]
See #2707
